overlord/devicestate/systems.go

```diff
@@ -176,3 +186,212 @@ func seededSystemFromModeenv() (*seededSystem, error) {
+	modelSnaps := make(map[string]*snap.Info)
+
+	getModelSnap := func(name string, essential bool, nonEssentialPresence string) error {
+		kind := "essential"
+		if !essential {
+			kind = "non-essential"
+			if nonEssentialPresence != "" {
+				kind = fmt.Sprintf("non-essential but %q",
+					nonEssentialPresence)
+			}
+		}
+		info, present, err := getInfo(name)
+		if err != nil {
+			return fmt.Errorf("cannot obtain %v snap information: %v", kind, err)
+		}
+		if !essential && !present && nonEssentialPresence == "optional" {
+			// non-essential snap which is declared as optionally
+			// present in the model
+			return nil
+		}
+		// grab those
+		logger.Noticef("%v snap: %v", kind, name)
+		if !present {
+			return fmt.Errorf("internal error: %v snap %q not present", kind, name)
+		}
+		if _, ok := modelSnaps[info.MountFile()]; ok {
+			// we've already seen this snap
+			return nil
+		}
+		// present locally
+		optsSnaps = append(optsSnaps, &seedwriter.OptionsSnap{
+			Path: info.MountFile(),
+		})
+		modelSnaps[info.MountFile()] = info
+		return nil
+	}
+
+	// snapd is implicitly needed
+	const snapdIsEssential = true
+	if err := getModelSnap("snapd", snapdIsEssential, ""); err != nil {
+		return nil, "", err
+	}
+	for _, sn := range model.EssentialSnaps() {
+		const essential = true
+		if err := getModelSnap(sn.SnapName(), essential, ""); err != nil {
+			return nil, "", err
+		}
+	}
+	for _, sn := range model.SnapsWithoutEssential() {
+		const essential = false
+		if err := getModelSnap(sn.SnapName(), essential, sn.Presence); err != nil {
+			return nil, "", err
+		}
+	}
+	if err := w.SetOptionsSnaps(optsSnaps); err != nil {
+		return nil, "", err
+	}
+
+	newFetcher := func(save func(asserts.Assertion) error) asserts.Fetcher {
+		fromDB := func(ref *asserts.Ref) (asserts.Assertion, error) {
+			return ref.Resolve(db.Find)
+		}
+		return asserts.NewFetcher(db, fromDB, save)
+	}
+	f, err := w.Start(db, newFetcher)
+	if err != nil {
+		return nil, "", err
+	}
+	// past this point the system directory is present
+
+	localSnaps, err := w.LocalSnaps()
+	if err != nil {
+		return nil, recoverySystemDir, err
+	}
+
+	for _, sn := range localSnaps {
+		info, ok := modelSnaps[sn.Path]
+		if !ok {
+			return nil, recoverySystemDir, fmt.Errorf("internal error: no snap info for %q", sn.Path)
+		}
+		// TODO: the side info derived here can be different from what
+		// we have in snap.Info, but getting it this way can be
+		// expensive as we need to compute the hash, try to find a
+		// better way
+		_, aRefs, err := seedwriter.DeriveSideInfo(sn.Path, f, db)
+		if err != nil {
+			if !asserts.IsNotFound(err) {
+				return nil, recoverySystemDir, err
+			} else if info.SnapID != "" {
+				// snap info from state must have come
+				// from the store, so it is unexpected
+				// if no assertions for it were found
+				return nil, recoverySystemDir, fmt.Errorf("internal error: no assertions for asserted snap with ID: %v", info.SnapID)
```

Comment:
thanks, there is indeed a check about this in seedwriter as well, but the error from there is more generic